py/persistentcode.c

```diff
@@ -164,51 +178,167 @@ STATIC void load_bytecode_qstrs(mp_reader_t *reader, byte *ip, byte *ip_top) {
+    size_t prelude_offset;
     bytecode_prelude_t prelude;
-    extract_prelude(&ip, &ip2, &prelude);
-
-    // load qstrs and link global qstr ids into bytecode
-    qstr simple_name = load_qstr(reader);
-    qstr source_file = load_qstr(reader);
-    ((byte*)ip2)[0] = simple_name; ((byte*)ip2)[1] = simple_name >> 8;
-    ((byte*)ip2)[2] = source_file; ((byte*)ip2)[3] = source_file >> 8;
-    load_bytecode_qstrs(reader, (byte*)ip, bytecode + bc_len);
-
-    // load constant table
-    size_t n_obj = read_uint(reader);
-    size_t n_raw_code = read_uint(reader);
-    mp_uint_t *const_table = m_new(mp_uint_t, prelude.n_pos_args + prelude.n_kwonly_args + n_obj + n_raw_code);
-    mp_uint_t *ct = const_table;
-    for (size_t i = 0; i < prelude.n_pos_args + prelude.n_kwonly_args; ++i) {
-        *ct++ = (mp_uint_t)MP_OBJ_NEW_QSTR(load_qstr(reader));
+    mp_uint_t type_sig = 0;
+    if (kind == MP_CODE_BYTECODE || kind == MP_CODE_NATIVE_PY) {
+        // extract prelude
+        if (kind == MP_CODE_NATIVE_PY) {
+            prelude_offset = read_uint(reader);
+            ip += prelude_offset;
+        }
+        const byte *ip2;
+        extract_prelude(&ip, &ip2, &prelude);
+
+        // load qstrs
+        qstr simple_name = load_qstr(reader);
+        qstr source_file = load_qstr(reader);
+        ((byte*)ip2)[0] = simple_name; ((byte*)ip2)[1] = simple_name >> 8;
+        ((byte*)ip2)[2] = source_file; ((byte*)ip2)[3] = source_file >> 8;
+    } else {
+        // load scope info
+        prelude.scope_flags = read_uint(reader);
+        prelude.n_pos_args = read_uint(reader);
+        prelude.n_kwonly_args = 0;
+        if (kind == MP_CODE_NATIVE_ASM) {
+            type_sig = read_uint(reader);
+        }
     }
-    for (size_t i = 0; i < n_obj; ++i) {
-        *ct++ = (mp_uint_t)load_obj(reader);
+
+    size_t n_qstr_link = 0;
+    if (kind == MP_CODE_BYTECODE) {
+        // link global qstr ids into bytecode
+        load_bytecode_qstrs(reader, (byte*)ip, fun_data + fun_data_len);
+    } else if (kind == MP_CODE_NATIVE_PY || kind == MP_CODE_NATIVE_VIPER) {
+        // load qstr link table
+        n_qstr_link = read_uint(reader);
+        for (size_t i = 0; i < n_qstr_link; ++i) {
+            qstr qst = load_qstr(reader);
+            size_t off = read_uint(reader);
+            link_qstr(fun_data + (off >> 1), off & 1, qst);
+        }
     }
-    for (size_t i = 0; i < n_raw_code; ++i) {
-        *ct++ = (mp_uint_t)(uintptr_t)load_raw_code(reader);
+
+    mp_uint_t *const_table = NULL;
+    if (kind != MP_CODE_NATIVE_ASM) {
+        // load constant table
+        size_t n_obj = read_uint(reader);
+        size_t n_raw_code = read_uint(reader);
+        const_table = m_new(mp_uint_t, prelude.n_pos_args + prelude.n_kwonly_args + n_obj + n_raw_code);
+        mp_uint_t *ct = const_table;
+        size_t i;
+        if (kind != MP_CODE_NATIVE_VIPER) {
+            for (i = 0; i < prelude.n_pos_args + prelude.n_kwonly_args; ++i) {
```

Comment:
I woulnd't find comments like "first entries in const table are function argument names" superfluous.

Reply:
Ok, I added a few comments to the load/save functions.